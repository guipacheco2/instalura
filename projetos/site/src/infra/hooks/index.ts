export * from './useForm'
export * from './usePostLike'
