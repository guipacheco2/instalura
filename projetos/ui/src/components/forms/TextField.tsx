import React from 'react'
import styled, { css } from 'styled-components'
import { Text, TextPropsGeneric } from '../foundation'

const StyledTextField = styled.div`
  margin-bottom: 17px;
`

const StyledInput = styled(Text)`
  border-radius: ${({ theme }) => theme.borderRadius};
  border: 1px solid ${({ theme }) => theme.colors.tertiary.light.color};
  outline: 0;
  padding: 12px 16px;
  width: 100%;

  ${({ isFieldInvalid, theme }) =>
    isFieldInvalid &&
    css`
      border-color: ${theme.colors.error.main.color};
      & + span {
        color: ${theme.colors.error.main.color};
        font-size: 11px;
      }
    `}
` as React.ComponentType<
  TextPropsGeneric<'input'> & { isFieldInvalid: boolean }
>

export interface TextFieldProps {
  error?: string
  isTouched?: boolean
  name: HTMLInputElement['name']
  onBlur?: (event: React.FocusEvent<HTMLInputElement>) => void
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void
  placeholder: HTMLInputElement['placeholder']
  type?: HTMLInputElement['type']
  value?: HTMLInputElement['value']
}

export function TextField({
  error = '',
  isTouched = false,
  name,
  onBlur,
  onChange,
  placeholder,
  type = 'text',
  value,
}: TextFieldProps): JSX.Element {
  const hasError = Boolean(error)
  const isFieldInvalid = hasError && isTouched

  return (
    <StyledTextField>
      <StyledInput
        as="input"
        placeholder={placeholder}
        name={name}
        type={type}
        onBlur={onBlur}
        onChange={onChange}
        value={value}
        variant="paragraph1"
        isFieldInvalid={isFieldInvalid}
      />

      {isFieldInvalid && (
        <Text variant="smallestException" color="error.main" role="alert">
          {error}
        </Text>
      )}
    </StyledTextField>
  )
}
