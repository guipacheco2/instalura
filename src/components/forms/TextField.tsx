import React from 'react'
import styled, { css } from 'styled-components'
import { Text, TextPropsGeneric } from '../foundation'

const StyledTextField = styled.div`
  margin-bottom: 17px;
`

const StyledInput = styled(Text)`
  width: 100%;
  border: 1px solid ${({ theme }) => theme.colors.tertiary.light.color};
  padding: 12px 16px;
  outline: 0;
  border-radius: ${({ theme }) => theme.borderRadius};

  ${({ isFieldInvalid, theme }) =>
    isFieldInvalid &&
    css`
      border-color: ${theme.colors.error.main.color};
      & + span {
        color: ${theme.colors.error.main.color};
        font-size: 11px;
      }
    `}
` as React.ComponentType<
  TextPropsGeneric<'input'> & { isFieldInvalid: boolean }
>

interface TextFieldProps {
  error?: string
  isTouched?: boolean
  name: HTMLInputElement['name']
  onBlur?: (event: React.FocusEvent<HTMLInputElement>) => void
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void
  placeholder: HTMLInputElement['placeholder']
  type?: HTMLInputElement['type']
  value?: HTMLInputElement['value']
}

export function TextField({
  error = '',
  isTouched = false,
  name,
  onBlur,
  onChange,
  placeholder,
  type = 'text',
  value,
}: TextFieldProps): JSX.Element {
  const hasError = Boolean(error)
  const isFieldInvalid = hasError && isTouched

  return (
    <StyledTextField>
      <StyledInput
        as="input"
        placeholder={placeholder}
        name={name}
        type={type}
        onBlur={onBlur}
        onChange={onChange}
        value={value}
        variant="paragraph1"
        isFieldInvalid={isFieldInvalid}
      />

      {isFieldInvalid && (
        <Text variant="smallestException" color="error.main" role="alert">
          {error}
        </Text>
      )}
    </StyledTextField>
  )
}
