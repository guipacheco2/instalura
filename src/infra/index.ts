export * from './cms'
export * from './env'
export * from './hooks'
export * from './http'
export * from './test'
