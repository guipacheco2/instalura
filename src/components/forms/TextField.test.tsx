/// <reference types="jest" />

import user from '@testing-library/user-event'
import React from 'react'
import { render, screen } from '../../infra'
import { TextField } from './TextField'

describe('<TextField />', () => {
  test('renders component', () => {
    render(
      <TextField
        placeholder="Nome"
        value="Ju"
        onChange={() => {
          console.log('input changed')
        }}
        name="nome"
      />,
    )

    const textField = screen.getByPlaceholderText(/nome/i)
    expect(textField).toMatchSnapshot()
  })

  describe('when field is valid', () => {
    describe('and user is typing', () => {
      test('the value must be updated', () => {
        const onChangeMock = jest.fn()
        render(
          <TextField
            placeholder="Nome"
            value=""
            onChange={onChangeMock}
            name="nome"
            isTouched
          />,
        )

        const inputNome = screen.getByPlaceholderText(/nome/i)
        user.type(inputNome, 'devsoutinho@gmail.com')
        expect(onChangeMock).toHaveBeenCalledTimes(21)
      })
    })
  })

  describe('when field is invalid', () => {
    test('displays the respective error message', () => {
      render(
        <TextField
          placeholder="Email"
          value="devsoutinhogmail.com"
          onChange={() => false}
          name="email"
          isTouched
          error="O campo email é obrigatório"
        />,
      )

      const inputEmail = screen.getByPlaceholderText(/email/i)
      expect(inputEmail).toHaveValue('devsoutinhogmail.com')
      expect(screen.getByRole('alert')).toHaveTextContent(
        'O campo email é obrigatório',
      )
      expect(inputEmail).toMatchSnapshot()
    })
  })
})
