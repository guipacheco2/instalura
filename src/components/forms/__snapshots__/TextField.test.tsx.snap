// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> renders component 1`] = `
.c0 {
  border-radius: 12px;
  border: 1px solid #88989E;
  outline: 0;
  padding: 12px 16px;
  width: 100%;
}

<input
  class="c0"
  name="nome"
  placeholder="Nome"
  type="text"
  value="Ju"
/>
`;

exports[`<TextField /> when field is invalid displays the respective error message 1`] = `
.c0 {
  border-radius: 12px;
  border: 1px solid #88989E;
  outline: 0;
  padding: 12px 16px;
  width: 100%;
  border-color: #dc3545;
}

.c0 + span {
  color: #dc3545;
  font-size: 11px;
}

<input
  class="c0"
  name="email"
  placeholder="Email"
  type="text"
  value="someusername"
/>
`;
